api/src/main/java/io/minio/CopySource.java

```diff
@@ -0,0 +1,136 @@
+import io.minio.errors.InvalidArgumentException;
+
+/** SourceInfo creates the SourceObject
+ * used for Compose Object.
+ */
+public class CopySource {
+  private String bucketName;
+  private String objectName;
+  private Integer offset;
+  private Integer length;
+  private ServerSideEncryption sseSource;
+  private Map<String, String> headerMap = new HashMap<>();
+
+  /**
+   * Constucts the CopySource with bucket name and object name.
+   */
+  public CopySource(String bucketName, String objectName ) {
+    this.bucketName = bucketName;
+    this.objectName = objectName;
+    this.offset = 0;
+    this.length = 0;
+  }
+  /**
+   * Constucts the CopySource with bucket name, object name, offset and length .
+   */
+  public CopySource(String bucketName, String objectName, Integer offset, Integer length) {
+    this.bucketName = bucketName;
+    this.objectName = objectName;
+    this.offset = offset;
+    this.length = length;
+  }
+
+  /**
+   * Constucts the CopySource with bucket name, object name, offset, length and headerMap .
+   */
+  public CopySource(String bucketName, String objectName, Integer offset, Integer length,
+                    Map<String, String> headerMap) {
+    this.bucketName = bucketName;
+    this.objectName = objectName;
+    this.offset = offset;
+    this.length = length;
+    this.headerMap = headerMap;
+  }
+
+
+  /**
+   * Constucts the CopySource with bucket name, object name, offset, length, headerMap and Server Side encryption .
+   */
+  public CopySource(String bucketName, String objectName, Integer offset, Integer length,
+                    ServerSideEncryption sseSource, Map<String, String> headerMap) {
+    this.bucketName = bucketName;
+    this.objectName = objectName;
+    this.offset = offset;
+    this.length = length;
+    this.headerMap = headerMap;
+    this.sseSource = sseSource;
+  }
+
+  /**
+   * Returns bucket name.
+   */
+  public String bucketName() {
+    return bucketName;
+  }
+
+  /**
+   * Returns object name.
+   */
+  public String objectName() {
+    return objectName;
+  }
+
+
+  /**
+   * Returns length.
+   */
+  public int length() throws InvalidArgumentException {
+    if ( length == null ) {
+      length = 0;
+    }
+    if ( offset == null ) {
+      offset = 0;
+    }
+    int size = length - offset;
+    if ( size < 0 ) {
+      throw new InvalidArgumentException("offset can not be greater the length");
+    }
+    return length - offset;
+  }
+
+  /**
+   * Returns the headerMap after setting copy-source, copy-source-range .
+   */
+  public Map<String, String> headers() throws InvalidArgumentException {
+/*    if (offset != -1) {
+      headerMap.put("x-amz-copy-source-range", "bytes=" + offset + "-" + length);
+    }*/
+
+
+    if (offset == null) {
+      if ( length != null) {
+        headerMap.put("x-amz-copy-source-range", "bytes="  + "0-" + length);
+      } else {
+        throw new InvalidArgumentException("x-amz-copy-source-range header is not set.");
```

Comment:
Is it required to set `x-amz-copy-source-range` for put object copy?

Reply:
yes.